pkg/vsphere/tasks/waiter.go

```diff
@@ -278,6 +278,41 @@ func IsNotFoundError(err error) bool {
 	return false
 }
 
+// IsInvalidPowerStateError is an error certifier function for errors coming back from vsphere. It checks for an InvalidPowerStateFault
+func IsInvalidPowerStateError(err error) bool {
+	if soap.IsVimFault(err) {
+		_, ok1 := soap.ToVimFault(err).(*types.InvalidPowerState)
+		_, ok2 := soap.ToVimFault(err).(*types.InvalidPowerStateFault)
+		return ok1 || ok2
+	}
+
+	if soap.IsSoapFault(err) {
+		_, ok1 := soap.ToSoapFault(err).VimFault().(types.InvalidPowerState)
+		_, ok2 := soap.ToSoapFault(err).VimFault().(types.InvalidPowerStateFault)
+		// sometimes we get the correct fault but wrong type
+		return ok1 || ok2 || soap.ToSoapFault(err).String == "vim.fault.InvalidPowerState" ||
+			soap.ToSoapFault(err).String == "vim.fault.InvalidPowerState"
+	}
+	return false
```

Comment:
This whole method seems to be basically repeated many times. Would there be value in defining a helper that takes the four "variable" pieces (two interfaces, two strings) as arguments? (If so, that should probably be handled in a separate PR.)

Reply:
The entire thing should be in govmomi. I think it would end up being either significantly rewritten to use reflection (viable but much harder to read) or would take four types as the pointers are separate types in this case.

I'd prefer just to move it all to govmomi and make it @dougm's problem ;)